keps/sig-network/1880-services-ip-ranges/README.md

```diff
@@ -0,0 +1,885 @@
+requiring complex and "unsupported" operations.
+
+#### Story 3
+
+As a Kubernetes developer I want to be able to know the current Service IP range and the IP addresses allocated.
+
+#### Story 4
+
+As a Kubernetes admin that wants to use IPv6, I want to be able to follow the IPv6 address
+planning best practices, being able to assign /64 prefixes to my end subnets.
+
+#### Story 5
+
+As a Kubernetes admin using a dual-stack cluster I want to be able to change the primary IP family of the cluster.
+
+### Notes/Constraints/Caveats (Optional)
+
+Current API machinery doesn't consider transactions, Services API is kind of special in this regard since already
+performs allocations inside the Registry pipeline using the Storage to keep consistency on an Obscure API Object that
+stores the bitmap with the allocations.
+
+This proposal also requires to perform validation from one Object against other Object inside an API request.
+
+### Risks and Mitigations
+
+Cross validation of Objects is not common in the Kubernetes API, sig-apimachinery should verify not bad precedent is introduced.
+
+Current Kubernetes cluster have a very static network configuration, exposing the Service IP range configuration will give
+more flexibility to users, with the risk to have problematic or incongruent network configurations.
+
+The Pod Network configuration is not really a Cluster configuration option, some CNI plugins use the node.Spec.ClusterCIDR field,
+but that is not guaranteed to be used. There is an effort in keps/sig-network/2593-multiple-cluster-cidrs to expose this configuration,
+but this also presents other cross-validation problem.
+
+## Design Details
+
+### API
+
+Define two new API objects
+
+#### IPRange
+
+##### Type
+
+```go
+// IPRange defines a range of IPs using CIDR format (192.168.0.0/24 or 2001:db2::0/64).
+type IPRange struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   IPRangeSpec   `json:"spec,omitempty"`
+	Status IPRangeStatus `json:"status,omitempty"`
+}
+
+
+// IPRangeSpec describe how the IPRange's specification looks like.
+type IPRangeSpec struct {
+	// An IPv4 block in CIDR notation "10.0.0.0/8"
+	IPv4 string `json:"ipv4"`
+
+	// An IPv6 block in CIDR notation "fd12:3456:789a:1::/64"
+	IPv6 string `json:"ipv6"`
+
+  // Users may reserve the first number of IPs in each range to be not used for automatic allocation.
+  ReservedSize int
+
+  // Primary defines the primary IP family of the IP range (default to IPv4)
+  Primary enum(ipv4|ipv6)
+}
+
+// IPRangeStatus indicates if the IPRange is ready to be used for allocating IPs
+// If not ready not IPAddress object can be 
+// xref: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties
+type IPRangeStatus struct {
+  // Conditions holds an array of metav1.Condition that describe the state of the IPRange.
+  Conditions []metav1.Conditon
```

Comment:
The idea of using a Status field is to signal the allocators on the apiservers, so they can know when they can use the Range to allocate IP.

This will centralize the operations on IP ranges on the controller, and will allow use to control complex operations like cross validation on creation and handle the delete process, or just temporary make will range unavailable to perform operations like renumbering (moving services from one range to other) of services safely ...

Reply:
removed the status,